express_server.js

```diff
@@ -6,135 +6,175 @@ const cookieSession = require("cookie-session");
 //renders register template
-app.get('/register', function(req, res) {
-  let templateVars = { urls: urlDatabase, user:users[req.session.user_id] };
+app.get("/register", (req, res) => {
+  const templateVars = { user: req.session.user_id };
   res.render("user_registration", templateVars);
 });
 
-app.post('/login', function(req, res) {
-  if (!req.body.email || !req.body.password) {
-    // errors if fields are empty
-    res.status(400).send("Oops! Please enter an email and password :) <a href='/login'>Login</a>");
-  } else if (getUserByEmail(req.body.email, users) === undefined) {
-    // errors if user is not registered
-    res.status(403).send("Looks like we're not friends yet :( This email does not exist, please <a href='/register'>register!</a>");
-  } else if (getUserByEmail(req.body.email, users) && !bcrypt.compareSync(req.body.password, users[getUserByEmail(req.body.email, users)].password)) {
-    // errors if user is registered and passwords do not match
-    res.status(403).send("Oops! Password and email do not match! <a href='/login'>Login</a>");
-  } else if (getUserByEmail(req.body.email, users) && bcrypt.compareSync(req.body.password, users[getUserByEmail(req.body.email, users)].password)) {
+app.post("/login", (req, res) => {
+  const email = req.body.email;
+  const password = req.body.password;
+  if (
+    emailExists(users, email) &&
+    areValidEmailPassword(users, email, password)
+  ) {
+    // set the cookie
     req.session.user_id = getUserByEmail(req.body.email, users);
     res.redirect("/urls");
+  } else {
+    res.status(400).redirect("/errors/400");
   }
 });
 
 //renders the user log-in page
-app.get('/login/', function(req, res) {
-  let templateVars = { urls: urlDatabase, user:users[req.session.user_id] };
+app.get("/login/", (req, res) => {
+  let templateVars = { user: req.session.user_id };
   res.render("user_login", templateVars);
 });
 
-app.get("/logout", function(req, res) {
+app.get("/logout", (req, res) => {
   req.session = null;
-  res.redirect('/urls');
+  res.redirect("/login");
 });
 
 // main page
 app.get("/", (req, res) => {
-  let templateVars = { urls: userURLs(req.session.user_id, urlDatabase), user: users[req.session.user_id] };
-  if (templateVars.user === undefined) {
-    res.redirect("/login");
-  } else {
-    res.redirect("/urls");
-  }
+  res.redirect("/urls");
 });
 
 // lists all the existing short URLs saved to the database
 app.get("/urls", (req, res) => {
-  let templateVars = { urls: userURLs(req.session.user_id, urlDatabase), user: users[req.session.user_id] };
-  if (templateVars.user === undefined) {
-    res.status(403).send("Tiny Dumpling does not know who you are! Please <a href='/login'>log in</a>!");
-  } else {
-    res.render("urls_index", templateVars);
-  }
+  const user_id = req.session.user_id;
+  !user_id
+    ? res.redirect("/login")
+    : res.render("urls_index", {
+        urls: userURLs(user_id, urlDatabase),
+        user: users[user_id],
+      });
```

Comment:
This is another new feature to make code less verbose.
It's a ternary condition, there are some caveats in order to use it (compared to the good old if..else) but it's great when you have situations like these (usually an if and only one else) because it makes the code very straight forward.
condition ? do this : or this.